torchaudio_contrib/layers.py

```diff
@@ -1,142 +1,241 @@
-
-
 import torch
+import math
 import torch.nn as nn
-import torch.nn.functional as F
 
-from .functional import spectrogram, create_mel_filter
+from .functional import stft_defaults, _stft, complex_norm, \
+    create_mel_filter, phase_vocoder
 
 
-class Spectrogram(nn.Module):
+class _ModuleNoStateBuffers(nn.Module):
     """
-    Module that outputs the spectrogram
-    of an audio signal with shape (batch, channel, time_hop, frequency_bins).
-
-    Its implemented as a layer so that the computation can be faster (done dynamically
-    on GPU) and no need to store the transforms. More information:
-        - https://github.com/keunwoochoi/kapre
-        - https://arxiv.org/pdf/1706.05781.pdf
+    Extension of nn.Module that removes buffers
+    from state_dict.
 
     Args:
-     * hop: int > 0
-       -  Hop length between frames in sample,  should be <= n_fft.
-       -  Default: None (in which case n_fft // 4 is used)
-     * n_fft: int > 0 
-       -  Size of the fft.
-       -  Default: 2048
-     * pad: int >= 0
-       -  Amount of two sided zero padding to apply.
-       -  Default: 0
-     * window: torch.Tensor,
-       -  Windowing used in the stft.
-       -  Default: None (in which case torch.hann_window(n_fft) is used)
-     * sr: int > 0
-       -  Sampling rate of the audio signal. This may not be the same in all samples (?)
-       -  Default: 44100
-     * power: int = 1,2,None
-       -  Power to normalize the spectrogram to, make None to work with complex stft
-       -  Default: 1
-     * kwargs: 
-       -  Any named arguments to be passed to the stft
+        buffers (dict): mapping of names and tensors to be registered.
+    """
 
+    def __init__(self, buffers):
+        super(_ModuleNoStateBuffers, self).__init__()
+        for k, v in buffers.items():
+            self.register_buffer(k, v)
```

Comment:
What is this good for? Can't the descendants do this as usual?

Reply:
Yeah definitely, I just figured that since `_ModuleNoStateBuffers` was responsible for removing the buffers from the `state_dict` it should also be creating them? But I guess that's not necessarily true.

Reply:
Yes, if it unconditionally removes all the buffers, it doesn't matter who created them. If we want descendants to be able to define which buffers to remove, then there's a point for setting these keys somewhere, but doing this in the constructor may not always be convenient. (But actually this should all be done by PyTorch, see https://github.com/pytorch/pytorch/issues/18056.)

Reply:
Yeah fair enough, I'll take the `register_buffer` out of the constructor.

>(But actually this should all be done by PyTorch, see pytorch/pytorch#18056.)

:+1: